src/backend/distributed/utils/enable_ssl.c

```diff
@@ -0,0 +1,444 @@
+static bool CreateCertificatesWhenNeeded(void);
+static EVP_PKEY * GeneratePrivateKey(void);
+static X509 * CreateCertificate(EVP_PKEY *privateKey);
+static bool StoreCertificate(EVP_PKEY *privateKey, X509 *certificate);
+#endif /* USE_SSL */
+
+
+PG_FUNCTION_INFO_V1(citus_setup_ssl);
+PG_FUNCTION_INFO_V1(citus_check_defaults_for_sslmode);
+
+
+/*
+ * citus_setup_ssl is called during the first creation of a citus extension. It configures
+ * postgres to use ssl if not already on. During this process it will create certificates
+ * if they are not already installed in the configured location.
+ */
+Datum
+citus_setup_ssl(PG_FUNCTION_ARGS)
+{
+#ifndef USE_SSL
+	ereport(WARNING, (errmsg("can not setup ssl on postgres that is not compiled with "
+							 "ssl support")));
+#else /* USE_SSL */
+	if (!EnableSSL && ShouldUseAutoSSL())
+	{
+		Node *enableSSLParseTree = NULL;
+
+		ereport(LOG, (errmsg("citus extension created on postgres without ssl enabled, "
+							 "turning it on during creation of the extension")));
+
+		/* execute the alter system statement to enable ssl on within postgres */
+		enableSSLParseTree = ParseTreeNode(ENABLE_SSL_QUERY);
+		AlterSystemSetConfigFile((AlterSystemStmt *) enableSSLParseTree);
+
+		/*
+		 * ssl=on requires that a key and certificate are present, since we have
+		 * enabled ssl mode here chances are the user didn't install credentials already.
+		 *
+		 * This function will check if they are available and if not it will generate a
+		 * self singed certificate.
+		 */
+		CreateCertificatesWhenNeeded();
+
+#if PG_VERSION_NUM >= 100000
+		GloballyReloadConfig();
+#else /* PG_VERSION_NUM < 100000 */
+		ereport(WARNING, (errmsg("restart of postgres required"),
+						  errdetail("citus enables ssl in postgres. Postgres "
+									"versions before 10.0 require a restart for changes "
+									"to ssl to take effect."),
+						  errhint("when restarting is not possible disable ssl and "
+								  "change citus.conn_nodeinfo to have sslmode lower "
+								  "then require.")));
+#endif
+	}
+#endif /* USE_SSL */
+
+	PG_RETURN_NULL();
+}
+
+
+/*
+ * citus_check_defaults_for_sslmode is called in the extension upgrade path when
+ * users upgrade from a previous version to a version that has ssl enabled by default, and
+ * only when the changed default value conflicts with the setup of the user.
+ *
+ * Once it is determined that the default value for citus.node_conninfo is used verbatim
+ * with ssl not enabled on the cluster it will reinstate the old default value for
+ * citus.node_conninfo.
+ *
+ * In effect this is to not impose the overhead of ssl on an already existing cluster that
+ * didn't have it enabled already.
+ */
+Datum
+citus_check_defaults_for_sslmode(PG_FUNCTION_ARGS)
+{
+	bool configChanged = false;
+
+	if (EnableSSL)
+	{
+		/* since ssl is on we do not have to change any sslmode back to prefer */
+		PG_RETURN_NULL();
+	}
+
+	if (strcmp(NodeConninfo, "sslmode=require") == 0)
```

Comment:
Might be worth adding a comment that this is only matches the exact value of the default, or maybe use constant in both places.

Reply:
I have thinking about a constant, but someone might then change the constant when we want to change the default again which would break this.

I will add a comment.